File: lib/screens/redesign/redesign_home.dart
// Home — the departures-first main view. A white header (nearest stop + crowd
// chip) over a scrolling sheet: a transfer-to-MRT card, then the live-arrivals
// list with a Time / Bus-number sort and an expandable "see all".

import 'dart:async';

import 'package:flutter/widgets.dart';
import 'package:material_symbols_icons/symbols.dart';

import '../../data/data_store.dart';
import '../../data/models.dart';
import '../../data/mrt_stations.dart';
import '../../services/location_service.dart';
import 'redesign_bridge.dart';
import 'redesign_common.dart';
import 'redesign_controller.dart';
import 'redesign_data.dart';
import 'redesign_theme.dart';

class RdHomeScreen extends StatefulWidget {
  const RdHomeScreen({super.key, required this.c});
  final RedesignController c;

  @override
  State<RdHomeScreen> createState() => _RdHomeScreenState();
}

class _RdHomeScreenState extends State<RdHomeScreen> {
  Timer? _pump;

  @override
  void initState() {
    super.initState();
    LocationService.shared.startIfAuthorized();
    DataStore.shared.prefetchNearbyArrivals();
    // Keep nearby arrivals warm (the freshness gate inside ensureArrivals
    // prevents over-fetching against LTA's rate limit).
    _pump = Timer.periodic(const Duration(seconds: 12), (_) {
      DataStore.shared.prefetchNearbyArrivals();
      final code = widget.c.currentNearby?.stopCode;
      if (code != null) DataStore.shared.ensureArrivals(code);
    });
  }

  @override
  void dispose() {
    _pump?.cancel();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    return Container(
      color: t.surface,
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.stretch,
        children: [
          _Header(c: widget.c),
          Expanded(child: _Sheet(c: widget.c)),
        ],
      ),
    );
  }
}

class _Header extends StatelessWidget {
  const _Header({required this.c});
  final RedesignController c;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    final stop = c.currentStop;
    return Padding(
      padding: const EdgeInsets.fromLTRB(18, 8, 18, 0),
      child: Column(
        children: [
          Row(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              Expanded(
                child: GestureDetector(
                  onTap: () => c.go('switch'),
                  behavior: HitTestBehavior.opaque,
                  child: Padding(
                    padding: const EdgeInsets.only(top: 6),
                    child: Column(
                      crossAxisAlignment: CrossAxisAlignment.start,
                      children: [
                        Row(
                          children: [
                            RdIcon(Symbols.near_me, size: 15, color: t.primary, fill: 1),
                            const SizedBox(width: 6),
                            Flexible(
                              child: Text(
                                  stop.distShort.isEmpty ? 'Nearest stop' : 'Nearest stop · ${stop.distShort}',
                                  maxLines: 1,
                                  overflow: TextOverflow.ellipsis,
                                  style: rdText(size: 12, weight: FontWeight.w600, color: t.onVariant)),
                            ),
                          ],
                        ),
                        const SizedBox(height: 3),
                        Row(
                          children: [
                            Flexible(
                              child: Text(stop.name,
                                  maxLines: 1,
                                  overflow: TextOverflow.ellipsis,
                                  style: rdText(
                                      size: 28, weight: FontWeight.w800, color: t.onSurface, letterSpacing: -0.7)),
                            ),
                            RdMrtBadgeRow(stopName: stop.name),
                            RdIcon(Symbols.chevron_right, size: 24, color: t.primary),
                          ],
                        ),
                      ],
                    ),
                  ),
                ),
              ),
              const SizedBox(width: 12),
              Padding(
                padding: const EdgeInsets.only(top: 2),
                child: Row(
                  children: [
                    Listener(
                      onPointerDown: (_) => c.saveDown(),
                      onPointerUp: (_) => c.saveUp(),
                      child: Container(
                        width: 42,
                        height: 42,
                        decoration: BoxDecoration(
                          shape: BoxShape.circle,
                          border: Border.all(color: t.outlineVariant),
                        ),
                        alignment: Alignment.center,
                        child: RdIcon(
                          c.stopSaved ? Symbols.bookmark : Symbols.bookmark_border,
                          size: 21,
                          color: c.stopSaved ? t.primary : t.onVariant,
                          fill: c.stopSaved ? 1 : 0,
                        ),
                      ),
                    ),
                    const SizedBox(width: 9),
                    RdCircleButton(
                      icon: Symbols.account_circle,
                      iconColor: t.onVariant,
                      onTap: () => c.go('settings'),
                    ),
                  ],
                ),
              ),
            ],
          ),
          Container(
            margin: const EdgeInsets.only(top: 14),
            padding: const EdgeInsets.only(bottom: 14),
            decoration: BoxDecoration(
              border: Border(bottom: BorderSide(color: t.outlineVariant)),
            ),
            child: Row(
              children: [
                Container(
                  padding: const EdgeInsets.symmetric(horizontal: 13, vertical: 8),
                  decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(13),
                    border: Border.all(color: t.outlineVariant),
                  ),
                  child: Row(
                    mainAxisSize: MainAxisSize.min,
                    children: [
                      RdIcon(Symbols.signpost, size: 16, color: t.onVariant, fill: 1),
                      const SizedBox(width: 7),
                      Text(stop.code.isEmpty ? stop.dist : 'Stop ${stop.code}',
                          style: rdText(size: 12.5, weight: FontWeight.w700, color: t.onSurface)),
                    ],
                  ),
                ),
              ],
            ),
          ),
        ],
      ),
    );
  }
}

class _Sheet extends StatelessWidget {
  const _Sheet({required this.c});
  final RedesignController c;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    final arrivals = c.visibleArrivals;
    return ListView(
      padding: const EdgeInsets.only(top: 4, bottom: 12),
      physics: const ClampingScrollPhysics(),
      children: [
        // Transfer-to-MRT card (order:-1 in the design → top of the sheet).
        _TransferCard(c: c),
        // LIVE ARRIVALS header.
        Padding(
          padding: const EdgeInsets.fromLTRB(18, 6, 18, 4),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: [
              Text('LIVE ARRIVALS',
                  style: rdText(size: 11, weight: FontWeight.w800, color: t.onVariant, letterSpacing: 0.66)),
              Row(
                children: [
                  RdDot(t.bus),
                  const SizedBox(width: 6),
                  Text(_freshnessLabel(c), style: rdText(size: 11, weight: FontWeight.w600, color: t.onVariant)),
                ],
              ),
            ],
          ),
        ),
        // Sort toggle.
        Padding(
          padding: const EdgeInsets.fromLTRB(18, 2, 18, 10),
          child: Align(
            alignment: Alignment.centerLeft,
            child: Container(
              padding: const EdgeInsets.all(3),
              decoration: BoxDecoration(color: t.scHigh, borderRadius: BorderRadius.circular(12)),
              child: Row(
                mainAxisSize: MainAxisSize.min,
                children: [
                  _SortTab(label: 'Time', active: c.sortBy == 'eta', onTap: () => c.setSort('eta')),
                  _SortTab(label: 'Bus number', active: c.sortBy == 'number', onTap: () => c.setSort('number')),
                ],
              ),
            ),
          ),
        ),
        // Arrivals.
        Padding(
          padding: const EdgeInsets.fromLTRB(12, 0, 12, 4),
          child: Column(
            children: [
              for (var i = 0; i < arrivals.length; i++)
                _ArrivalRow(c: c, a: arrivals[i], highlighted: i == 0),
              if (c.canExpandArrivals)
                GestureDetector(
                  onTap: c.toggleArrivals,
                  behavior: HitTestBehavior.opaque,
                  child: Padding(
                    padding: const EdgeInsets.all(11),
                    child: Row(
                      mainAxisAlignment: MainAxisAlignment.center,
                      children: [
                        Text(c.arrivalsExpanded ? 'Show fewer arrivals' : 'See all arrivals',
                            style: rdText(size: 12.5, weight: FontWeight.w700, color: t.primary)),
                        const SizedBox(width: 5),
                        RdIcon(c.arrivalsExpanded ? Symbols.expand_less : Symbols.expand_more,
                            size: 18, color: t.primary),
                      ],
                    ),
                  ),
                ),
            ],
          ),
        ),
      ],
    );
  }
}

class _SortTab extends StatelessWidget {
  const _SortTab({required this.label, required this.active, required this.onTap});
  final String label;
  final bool active;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    return GestureDetector(
      onTap: onTap,
      child: AnimatedContainer(
        duration: const Duration(milliseconds: 180),
        padding: const EdgeInsets.symmetric(horizontal: 18, vertical: 8),
        decoration: BoxDecoration(
          color: active ? t.primary : const Color(0x00000000),
          borderRadius: BorderRadius.circular(9),
        ),
        child: Text(label,
            style: rdText(size: 13, weight: FontWeight.w700, color: active ? t.onPrimary : t.onVariant)),
      ),
    );
  }
}

class _ArrivalRow extends StatelessWidget {
  const _ArrivalRow({required this.c, required this.a, required this.highlighted});
  final RedesignController c;
  final RdArrival a;
  final bool highlighted;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    final hi = highlighted;
    final occ = rdOcc(a.load, t);
    return GestureDetector(
      onTap: () => c.openBus(a.route, c.currentNearby?.stopCode),
      child: Container(
        margin: const EdgeInsets.only(bottom: 10),
        padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 13),
        decoration: BoxDecoration(
          color: hi ? t.primaryContainer : t.scLow,
          borderRadius: BorderRadius.circular(18),
          border: Border.all(color: hi ? t.primary : t.outlineVariant),
        ),
        child: Row(
          children: [
            Container(
              constraints: const BoxConstraints(minWidth: 50),
              height: 48,
              padding: const EdgeInsets.symmetric(horizontal: 6),
              alignment: Alignment.center,
              decoration: BoxDecoration(
                color: hi ? t.primary : t.scHigh,
                borderRadius: BorderRadius.circular(14),
              ),
              child: Text(a.route,
                  style: rdText(size: 19, weight: FontWeight.w800, color: hi ? t.onPrimary : t.onSurface)),
            ),
            const SizedBox(width: 13),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text(a.dest,
                      maxLines: 1,
                      overflow: TextOverflow.ellipsis,
                      style: rdText(
                          size: 16,
                          weight: FontWeight.w800,
                          color: hi ? t.onPrimaryContainer : t.onSurface,
                          letterSpacing: -0.16)),
                  const SizedBox(height: 5),
                  Row(
                    children: [
                      RdIcon(occ.icon, size: 16, color: occ.color, fill: 1),
                      const SizedBox(width: 5),
                      Flexible(
                        child: Text(occ.label,
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                            style: rdText(size: 11.5, weight: FontWeight.w600, color: occ.color)),
                      ),
                    ],
                  ),
                ],
              ),
            ),
            const SizedBox(width: 8),
            Column(
              crossAxisAlignment: CrossAxisAlignment.end,
              mainAxisSize: MainAxisSize.min,
              children: [
                Text.rich(
                  TextSpan(children: [
                    TextSpan(
                        text: a.min,
                        style: rdText(
                            size: 22,
                            weight: FontWeight.w900,
                            color: hi ? t.primary : t.onSurface,
                            height: 0.9,
                            letterSpacing: -0.44)),
                    TextSpan(text: ' min', style: rdText(size: 11, weight: FontWeight.w700, color: t.onVariant)),
                  ]),
                ),
                if (a.then != null) ...[
                  const SizedBox(height: 3),
                  Text(a.then!, style: rdText(size: 10.5, weight: FontWeight.w500, color: t.onVariant)),
                ],
              ],
            ),
            const SizedBox(width: 6),
            RdIcon(Symbols.chevron_right, size: 20, color: hi ? t.primary : t.outline),
          ],
        ),
      ),
    );
  }
}

String _freshnessLabel(RedesignController c) {
  final code = c.currentNearby?.stopCode;
  final last = code == null ? null : DataStore.shared.lastRefresh(code);
  if (last == null) return 'Updating…';
  final s = DateTime.now().difference(last).inSeconds;
  if (s < 5) return 'Updated just now';
  if (s < 60) return 'Updated ${s}s ago';
  return 'Updated ${s ~/ 60}m ago';
}

class _TransferCard extends StatelessWidget {
  const _TransferCard({required this.c});
  final RedesignController c;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    final nm = c.nearestMrt;
    if (nm == null) return const SizedBox.shrink();
    final st = nm.station;
    return GestureDetector(
      onTap: () => c.openStationNamed(st.name),
      child: Container(
        margin: const EdgeInsets.fromLTRB(14, 8, 14, 16),
        padding: const EdgeInsets.symmetric(horizontal: 14, vertical: 12),
        decoration: BoxDecoration(color: t.scHigh, borderRadius: BorderRadius.circular(18)),
        child: Row(
          children: [
            Container(
              width: 44,
              height: 44,
              decoration: BoxDecoration(color: t.transferOrange, borderRadius: BorderRadius.circular(13)),
              alignment: Alignment.center,
              child: RdIcon(Symbols.directions_subway, size: 23, color: t.transferOnOrange, fill: 1),
            ),
            const SizedBox(width: 12),
            Expanded(
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  Text('NEAREST MRT',
                      style: rdText(size: 9, weight: FontWeight.w700, color: t.onVariant, letterSpacing: 0.63)),
                  const SizedBox(height: 1),
                  Row(
                    children: [
                      Flexible(
                        child: Text(st.name,
                            maxLines: 1,
                            overflow: TextOverflow.ellipsis,
                            style: rdText(size: 16, weight: FontWeight.w800, color: t.onSurface, letterSpacing: -0.16)),
                      ),
                      for (final code in st.codes.take(2)) ...[
                        const SizedBox(width: 7),
                        Container(
                          padding: const EdgeInsets.symmetric(horizontal: 7, vertical: 2),
                          decoration: BoxDecoration(
                              color: lineColorFor(code), borderRadius: BorderRadius.circular(6)),
                          child: Text(code,
                              style: rdText(size: 9.5, weight: FontWeight.w800, color: rdMrtBadgeFg(code))),
                        ),
                      ],
                    ],
                  ),
                  const SizedBox(height: 2),
                  Text('${nm.walkMin} min walk · ${fmtDistance(nm.distanceM)}',
                      style: rdText(size: 11.5, weight: FontWeight.w500, color: t.onVariant)),
                ],
              ),
            ),
            const SizedBox(width: 8),
            Container(
              width: 38,
              height: 38,
              decoration: BoxDecoration(color: t.surface, shape: BoxShape.circle),
              alignment: Alignment.center,
              child: RdIcon(Symbols.directions_walk, size: 19, color: t.primary, fill: 1),
            ),
            const SizedBox(width: 4),
            RdIcon(Symbols.chevron_right, size: 20, color: t.outline),
          ],
        ),
      ),
    );
  }
}

/// Bottom Nearby / Saved navigation bar (top-level screens only).
class RdBottomNav extends StatelessWidget {
  const RdBottomNav({super.key, required this.c});
  final RedesignController c;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    final onMap = c.screen == 'map';
    return Container(
      decoration: BoxDecoration(
        color: t.surface,
        border: Border(top: BorderSide(color: t.outlineVariant)),
      ),
      child: SafeArea(
        top: false,
        child: Padding(
          padding: const EdgeInsets.fromLTRB(0, 7, 0, 5),
          child: Row(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            children: [
              _NavItem(
                label: 'Nearby',
                icon: Symbols.near_me,
                fill: 1,
                active: onMap,
                onTap: c.toMap,
              ),
              _NavItem(
                label: 'Saved',
                icon: Symbols.bookmark,
                fill: 0,
                active: false,
                onTap: () => c.go('saved'),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

class _NavItem extends StatelessWidget {
  const _NavItem({
    required this.label,
    required this.icon,
    required this.fill,
    required this.active,
    required this.onTap,
  });
  final String label;
  final IconData icon;
  final double fill;
  final bool active;
  final VoidCallback onTap;

  @override
  Widget build(BuildContext context) {
    final t = RdTheme.of(context);
    return GestureDetector(
      onTap: onTap,
      behavior: HitTestBehavior.opaque,
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          Container(
            width: 56,
            height: 28,
            alignment: Alignment.center,
            decoration: BoxDecoration(
              color: active ? t.primaryContainer : const Color(0x00000000),
              borderRadius: BorderRadius.circular(999),
            ),
            child: RdIcon(icon,
                size: active ? 20 : 21,
                color: active ? t.onPrimaryContainer : t.onVariant,
                fill: fill),
          ),
          const SizedBox(height: 2),
          Text(label,
              style: rdText(
                  size: 9.5,
                  weight: active ? FontWeight.w700 : FontWeight.w500,
                  color: active ? t.onSurface : t.onVariant)),
        ],
      ),
    );
  }
}

File: ios-native/Leyne/Redesign/RedesignHome.swift
// Home — the departures-first main view (iOS). A header (nearest stop + crowd
// chip) over a scrolling sheet: a transfer-to-MRT card, then the live-arrivals
// list with a Time / Bus-number sort and an expandable "see all".

import SwiftUI

struct RDHomeScreen: View {
    @ObservedObject var m: RedesignModel
    let t: RDTokens
    @EnvironmentObject private var store: DataStore
    @EnvironmentObject private var loc: LocationManager
    @State private var pump: Timer?
    @State private var requested: Set<String> = []

    var body: some View {
        VStack(spacing: 0) {
            header
            sheet
        }
        .background(t.surface)
        .onAppear {
            loc.startIfAuthorized()
            store.prefetchNearbyArrivals()
            startPump()
        }
        .onDisappear { stopPump() }
    }

    // Keeps the nearby stops' live arrivals warm — a few stops per tick so the
    // list stays current without bursting LTA's rate limit (mirrors NearbyView).
    private func startPump() {
        stopPump()
        pump = Timer.scheduledTimer(withTimeInterval: 0.8, repeats: true) { _ in
            Task { @MainActor in
                var fired = 0
                for s in store.nearby where fired < 3 {
                    if requested.insert(s.stopCode).inserted {
                        store.ensureArrivals(stop: s.stopCode, silent: true); fired += 1
                    }
                }
                if let code = m.currentNearby?.stopCode {
                    store.ensureArrivals(stop: code)
                }
            }
        }
    }

    private func stopPump() { pump?.invalidate(); pump = nil }

    // MARK: header

    private var header: some View {
        let stop = m.currentStop
        return VStack(alignment: .leading, spacing: 0) {
            HStack(alignment: .top, spacing: 12) {
                // Title block — plain text. Finding / switching stops is now an
                // explicit search button on the right, not a hidden tap on the
                // title (the chevron is gone).
                VStack(alignment: .leading, spacing: 3) {
                    Text("NEAREST STOP")
                        .font(rdFont(11, .bold)).kerning(0.7).foregroundStyle(t.onVariant)
                    HStack(spacing: 7) {
                        Text(stop.name)
                            .font(rdFont(30, .heavy)).foregroundStyle(t.onSurface)
                            .lineLimit(1).minimumScaleFactor(0.7)
                        RDMrtBadgeRow(stopName: stop.name)
                    }
                    Text(headerSubtitle)
                        .font(rdFont(13, .medium)).foregroundStyle(t.onVariant)
                        .lineLimit(1)
                }
                .frame(maxWidth: .infinity, alignment: .leading)

                // Lighter, toolbar-style actions — plain SF Symbols, no circle
                // chrome (the bordered circles read as heavy stacked together).
                HStack(spacing: 6) {
                    RDCircleButton(symbol: "magnifyingglass", bordered: false, iconColor: t.onVariant, t: t) { m.go("switch") }
                    saveButton
                    RDCircleButton(symbol: "person.crop.circle", bordered: false, iconColor: t.onVariant, t: t) { m.go("settings") }
                }
                .padding(.top, 4)
            }
            .padding(.horizontal, 18)
            .padding(.top, 8).padding(.bottom, 14)
            hairline
        }
    }

    /// Quiet supporting line under the stop title: distance + walk + code, folded
    /// into one caption (replaces the old bordered "Stop NNNNN" pill).
    private var headerSubtitle: String {
        guard let n = m.currentNearby, !m.currentStop.code.isEmpty else {
            return m.currentStop.dist   // e.g. "Waiting for your location"
        }
        if n.distanceM <= 40 { return "You're at this stop · Stop \(n.stopCode)" }
        return "\(fmtDistance(n.distanceM)) away · \(n.walkMin) min walk · Stop \(n.stopCode)"
    }

    // Thin dividers do the organising work instead of cards. A full-bleed line
    // separates sections; an inset line (aligned under the row text) separates
    // items within the arrivals list — the Apple Maps / Settings grouping idiom.
    private var hairline: some View { Rectangle().fill(t.outlineVariant).frame(height: 1) }
    private var insetHairline: some View {
        Rectangle().fill(t.outlineVariant).frame(height: 1).padding(.leading, 78)
    }

    private var saveButton: some View {
        RDSym(m.stopSaved ? "bookmark.fill" : "bookmark", size: 22, color: m.stopSaved ? t.primary : t.onVariant)
            .frame(width: 42, height: 42)
            .contentShape(Circle())
            .onTapGesture { m.toggleSaveStop() }
            .simultaneousGesture(LongPressGesture(minimumDuration: 0.45).onEnded { _ in
                m.notify("Opening saved stops"); m.go("saved")
            })
    }

    // MARK: sheet

    private var sheet: some View {
        ScrollView {
            LazyVStack(spacing: 0) {
                mrtRow
                liveSection
                nearbyStationsSection
            }
            .padding(.bottom, 16)
        }
    }

    // MARK: Nearby MRT stations — real, tappable context that fills the lower area

    @ViewBuilder private var nearbyStationsSection: some View {
        if let here = loc.location {
            let near = MrtGeo.nearestStations(to: here.coordinate, limit: 3, withinMeters: 1500)
            if !near.isEmpty {
                hairline
                HStack {
                    Text("NEARBY STATIONS").font(rdFont(11, .heavy)).kerning(0.66).foregroundStyle(t.onVariant)
                    Spacer()
                }
                .padding(.horizontal, 18).padding(.top, 16).padding(.bottom, 8)
                ForEach(Array(near.enumerated()), id: \.element.station.id) { idx, item in
                    if idx > 0 { insetHairline }
                    stationRow(item.station, walkMin: item.walkMin)
                }
            }
        }
    }

    private func stationRow(_ st: MrtGeoStation, walkMin: Int) -> some View {
        let lineColor = st.codes.first.map { mrtLineColorFor($0) } ?? t.transferOrange
        return Button(action: { m.openStation(named: st.name) }) {
            HStack(spacing: 14) {
                RDSym("tram.fill", size: 16, color: lineColor)
                    .frame(width: 46, height: 38)
                    .background(t.scHigh).clipShape(RoundedRectangle(cornerRadius: 11, style: .continuous))
                VStack(alignment: .leading, spacing: 3) {
                    Text(st.name).font(rdFont(16, .semibold)).foregroundStyle(t.onSurface).lineLimit(1)
                    HStack(spacing: 4) {
                        ForEach(st.codes.prefix(3), id: \.self) { c in
                            Text(c).font(rdFont(9, .heavy)).foregroundStyle(rdMrtBadgeFg(c))
                                .padding(.horizontal, 5).padding(.vertical, 1.5)
                                .background(mrtLineColorFor(c))
                                .clipShape(RoundedRectangle(cornerRadius: 5, style: .continuous))
                        }
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                Text("\(walkMin) min walk").font(rdFont(12.5, .medium)).foregroundStyle(t.onVariant)
            }
            .padding(.horizontal, 18).padding(.vertical, 13)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    // MARK: MRT — one quiet row, not a competing card

    @ViewBuilder private var mrtRow: some View {
        if let nm = m.nearestMrt {
            let st = nm.station
            let lineColor = st.codes.first.map { mrtLineColorFor($0) } ?? t.transferOrange
            let lineName = st.codes.first.flatMap { MRTLine(rawValue: String($0.prefix(2)).uppercased())?.displayName }
            Button(action: { m.openStation(named: st.name) }) {
                HStack(spacing: 11) {
                    // A thin line-coloured bar is the whole MRT accent.
                    Capsule().fill(lineColor).frame(width: 3.5, height: 30)
                    RDSym("tram.fill", size: 18, color: lineColor)
                    VStack(alignment: .leading, spacing: 1) {
                        Text(st.name).font(rdFont(15.5, .semibold)).foregroundStyle(t.onSurface).lineLimit(1)
                        Text(lineName.map { "\($0) Line" } ?? "MRT station")
                            .font(rdFont(11.5, .medium)).foregroundStyle(lineColor)
                    }
                    Spacer(minLength: 8)
                    Text("\(nm.walkMin) min walk").font(rdFont(12.5, .medium)).foregroundStyle(t.onVariant)
                    RDSym("chevron.right", size: 15, color: t.outline)
                }
                .padding(.horizontal, 18).padding(.vertical, 12)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)
            hairline
        }
    }

    // MARK: Live arrivals

    @ViewBuilder private var liveSection: some View {
        HStack(spacing: 7) {
            if isFresh { RDDot(color: t.bus) }
            Text("LIVE ARRIVALS").font(rdFont(11, .heavy)).kerning(0.66).foregroundStyle(t.onVariant)
            if !isFresh, let stale = staleLabel {
                Text("· \(stale)").font(rdFont(11, .medium)).foregroundStyle(t.onVariant)
            }
            Spacer()
            sortToggle
        }
        .padding(.horizontal, 18).padding(.top, 16).padding(.bottom, 8)

        if m.visibleArrivals.isEmpty {
            Text("No live arrivals right now")
                .font(rdFont(13.5, .medium)).foregroundStyle(t.onVariant)
                .frame(maxWidth: .infinity, alignment: .leading)
                .padding(.horizontal, 18).padding(.vertical, 22)
        } else {
            ForEach(Array(m.visibleArrivals.enumerated()), id: \.element.id) { idx, a in
                if idx > 0 { insetHairline }
                arrivalRow(a)
            }
            if m.canExpandArrivals {
                hairline
                showAllRow
            }
        }
    }

    /// Minimal text sort — no filled segmented surface. Active option in primary.
    private var sortToggle: some View {
        HStack(spacing: 12) {
            sortText("Time", "eta")
            sortText("Bus no.", "number")
        }
    }

    private func sortText(_ label: String, _ key: String) -> some View {
        let active = m.sortBy == key
        return Button(action: { withAnimation(.easeOut(duration: 0.18)) { m.setSort(key) } }) {
            Text(label).font(rdFont(12, active ? .bold : .medium))
                .foregroundStyle(active ? t.primary : t.onVariant)
        }
        .buttonStyle(.plain)
    }

    private var showAllRow: some View {
        Button(action: { withAnimation(.easeOut(duration: 0.25)) { m.toggleArrivals() } }) {
            HStack(spacing: 6) {
                Text(m.arrivalsExpanded ? "Show fewer" : "Show all \(m.sortedArrivals.count) arrivals")
                    .font(rdFont(13.5, .semibold)).foregroundStyle(t.primary)
                RDSym(m.arrivalsExpanded ? "chevron.up" : "chevron.down", size: 16, color: t.primary)
                Spacer()
            }
            .padding(.horizontal, 18).padding(.vertical, 15)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }

    // MARK: Freshness — "● Live" while recent; a quiet timestamp only once stale.

    private var freshnessSeconds: Int? {
        guard let code = m.currentNearby?.stopCode, let last = store.lastRefresh(code) else { return nil }
        return Int(Date().timeIntervalSince(last))
    }
    private var isFresh: Bool { (freshnessSeconds ?? 999) < 20 }
    private var staleLabel: String? {
        guard let s = freshnessSeconds else { return "Updating…" }
        if s < 60 { return "Updated \(s)s ago" }
        return "Updated \(s / 60)m ago"
    }

    /// "then 11" → "Next 11 min"; nil when there's no following bus.
    private func nextLabel(_ then: String?) -> String? {
        guard let then else { return nil }
        let digits = then.filter(\.isNumber)
        return digits.isEmpty ? nil : "Next \(digits) min"
    }

    /// One arrival as a flat list row (no card, no border, no chevron). The ETA
    /// dominates on the right; destination is secondary, occupancy tertiary. Every
    /// bus-number badge is the same neutral surface — urgency is carried by row
    /// order and the green "Now"/ETA, not by tinting one badge (blue reads as
    /// "selected" here, which it isn't).
    private func arrivalRow(_ a: RDArrival) -> some View {
        let occ = rdOcc(a.load, t)
        let arriving = (Int(a.min) ?? 99) <= 0
        return Button(action: { m.openBus(service: a.route, stopCode: m.currentNearby?.stopCode) }) {
            HStack(spacing: 14) {
                Text(a.route)
                    .font(rdFont(17, .heavy))
                    .foregroundStyle(t.onSurface)
                    .frame(minWidth: 46).frame(height: 38)
                    .padding(.horizontal, 8)
                    .background(t.scHigh)
                    .clipShape(RoundedRectangle(cornerRadius: 11, style: .continuous))
                VStack(alignment: .leading, spacing: 4) {
                    Text(a.dest).font(rdFont(16.5, .semibold)).foregroundStyle(t.onSurface).lineLimit(1)
                    HStack(spacing: 6) {
                        RDDot(color: occ.color, size: 7)
                        Text(occ.label).font(rdFont(12.5, .medium)).foregroundStyle(t.onVariant).lineLimit(1)
                    }
                }
                .frame(maxWidth: .infinity, alignment: .leading)
                VStack(alignment: .trailing, spacing: 2) {
                    if arriving {
                        Text("Now").font(rdFont(17, .heavy)).foregroundStyle(t.bus)
                    } else {
                        HStack(alignment: .firstTextBaseline, spacing: 2) {
                            Text(a.min).font(rdFont(24, .black)).foregroundStyle(t.onSurface)
                                .contentTransition(.numericText())   // cross-fades as it ticks
                            Text("min").font(rdFont(11, .bold)).foregroundStyle(t.onVariant)
                        }
                    }
                    if let next = nextLabel(a.then) {
                        Text(next).font(rdFont(10.5, .medium)).foregroundStyle(t.onVariant)
                    }
                }
                .animation(.easeInOut(duration: 0.35), value: a.min)
            }
            .padding(.horizontal, 18).padding(.vertical, 16)
            .contentShape(Rectangle())
        }
        .buttonStyle(.plain)
    }
}

// MARK: - Bottom nav

struct RDBottomNav: View {
    @ObservedObject var m: RedesignModel
    let t: RDTokens

    var body: some View {
        HStack {
            Spacer()
            navItem("Nearby", "location.north.fill", active: m.screen == "map") { m.toMap() }
            Spacer()
            navItem("Saved", "bookmark.fill", active: false) { m.go("saved") }
            Spacer()
        }
        .padding(.top, 7).padding(.bottom, 5)
        .background(t.surface)
        .overlay(Rectangle().fill(t.outlineVariant).frame(height: 1), alignment: .top)
    }

    private func navItem(_ label: String, _ symbol: String, active: Bool, _ tap: @escaping () -> Void) -> some View {
        Button(action: tap) {
            // Flat tab (iOS 26 direction): no capsule — the active tab is simply
            // tinted with the primary colour on both icon and label.
            VStack(spacing: 3) {
                RDSym(symbol, size: 22, color: active ? t.primary : t.onVariant)
                    .frame(height: 26)
                Text(label).font(rdFont(10, active ? .semibold : .medium))
                    .foregroundStyle(active ? t.primary : t.onVariant)
            }
        }
        .buttonStyle(.plain)
    }
}
